File: python/graph_search/a_star.py
'''
@file: a_star.py
@breif: A* motion planning
@author: Winter
@update: 2023.1.13
'''
import os, sys
import math
import heapq

sys.path.append(os.path.abspath(os.path.join(__file__, "../../")))

from .graph_search import GraphSearcher
from utils import Env, Node

class AStar(GraphSearcher):
    '''
    Class for A* motion planning.

    Parameters
    ----------
    start: tuple
        start point coordinate
    goal: tuple
        goal point coordinate
    env: Env
        environment
    heuristic_type: str
        heuristic function type, default is euclidean

    Examples
    ----------
    >>> from utils import Env
    >>> from graph_search import AStar
    >>> start = (5, 5)
    >>> goal = (45, 25)
    >>> env = Env(51, 31)
    >>> planner = AStar(start, goal, env)
    >>> planner.run()
    '''
    def __init__(self, start: tuple, goal: tuple, env: Env, heuristic_type: str = "euclidean") -> None:
        super().__init__(start, goal, env, heuristic_type)

    def __str__(self) -> str:
        return "A*"

    def plan(self):
        '''
        A* motion plan function.
        [1] A Formal Basis for the heuristic Determination of Minimum Cost Paths

        Return
        ----------
        path: list
            planning path
        expand: list
            all nodes that planner has searched
        '''
        # OPEN set with priority and CLOSED set
        OPEN = []
        heapq.heappush(OPEN, self.start)
        CLOSED = []

        while OPEN:
            node = heapq.heappop(OPEN)

            # exists in CLOSED set
            if node in CLOSED:
                continue

            # goal found
            if node == self.goal:
                CLOSED.append(node)
                return self.extractPath(CLOSED), CLOSED

            for node_n in self.getNeighbor(node):                
                # exists in CLOSED set
                if node_n in CLOSED:
                    continue
                
                node_n.parent = node.current
                node_n.h = self.h(node_n, self.goal)

                # goal found
                if node_n == self.goal:
                    heapq.heappush(OPEN, node_n)
                    break
                
                # update OPEN set
                heapq.heappush(OPEN, node_n)
            
            CLOSED.append(node)
        return [], []

    def getNeighbor(self, node: Node) -> list:
        '''
        Find neighbors of node.

        Parameters
        ----------
        node: Node
            current node

        Return
        ----------
        neighbors: list
            neighbors of current node
        '''
        return [node + motion for motion in self.motions
                if not self.isCollision(node, node + motion)]

    def extractPath(self, closed_set):
        '''
        Extract the path based on the CLOSED set.

        Parameters
        ----------
        closed_set: list
            CLOSED set

        Return
        ----------
        cost: float
            the cost of planning path
        path: list
            the planning path
        '''
        cost = 0
        node = closed_set[closed_set.index(self.goal)]
        path = [node.current]
        while node != self.start:
            index = closed_set.index(Node(node.parent, None, None, None))
            cost += math.hypot(node.current[0] - node.parent[0], node.current[1] - node.parent[1])
            node = closed_set[index]
            path.append(node.current)
        return cost, path

    def run(self):
        '''
        Running both plannig and animation.
        '''
        (cost, path), expand = self.plan()
        self.plot.animation(path, str(self), cost, expand)

File: matlab/graph_search/a_star.m
function [path, flag, cost, expand] = a_star(map, start, goal)
%%
% @file: a_star.m
% @breif: A* motion planning
% @paper: A Formal Basis for the heuristic Determination of Minimum Cost Paths
% @author: Winter
% @update: 2023.1.29

%{ 
    ============ OPEN and CLOSED==============
    [x, y, g, h, px, py]
    =======================================
%}

%%
% initialize
expand= [];                                                   % expand zone
OPEN  = [start, 0, h(start, goal), start];      % OPEN set with priority 
CLOSED = [];                                                % CLOSED set
flag  = false;
neighbor  = [-1, 1, 14;...  
                0, 1, 10;...
                1, 1, 14;...
              -1, 0, 10;...
                1, 0, 10;...
              -1, -1, 14;...
                0, -1, 10;...
                1, -1, 14];
neighbor_num = length(neighbor(:, 1));
cost = 0;

while ~isempty(OPEN(:, 1))
    % pop
    f = OPEN(:, 3) + OPEN(:, 4);
    [~, index] = min(f);
    cur_node = OPEN(index, :);
    OPEN(index, :) = []; 
    
    % exists in CLOSED set
    if loc_list(cur_node, CLOSED, [1, 2])
        continue
    end
    
    % goal found
    if cur_node(1) == goal(1) && cur_node(2) == goal(2)
        CLOSED = [cur_node; CLOSED];
        flag = true;
        cost = cur_node(3);
        break
    end
    
    % explore neighbors
    for i=1:neighbor_num        
        node_n = [cur_node(1) + neighbor(i, 1), ...
                          cur_node(2) + neighbor(i, 2), ...
                          cur_node(3) + neighbor(i, 3), ...
                          0, cur_node(1), cur_node(2)
                          ];
        node_n(4) = h(node_n(1:2), goal);
        
        % exists in CLOSED set
        if loc_list(cur_node, CLOSED, [1, 2])
            continue
        end
        
        % obstacle
        if map(node_n(1), node_n(2)) == 2
            continue;
        end
       
        % goal found
        if cur_node(1) == goal(1) && cur_node(2) == goal(2)
            CLOSED = [cur_node; CLOSED];
            flag = true;
            cost = cur_node(3);
            break
        end
        
        % update expand zone
        expand = [expand; node_n(1:2)];
        
        % update OPEN set
        OPEN = [OPEN; node_n];
    end
    CLOSED = [cur_node; CLOSED];
end

% extract path
path = extract_path(CLOSED, start);
end

%%
function h_val = h(cur_node, goal)
% @breif: heuristic function(Manhattan distance)
    h_val = 10 * abs(cur_node(1) - goal(1)) + 10 * abs(cur_node(2) - goal(2));
end

function index = loc_list(node, list, range)
% @breif: locate the node in given list
    num = size(list);
    index = 0;
    if ~num(1)
        return
    else  
        for i=1:num(1)
            if isequal(node(range), list(i, range))
                index = i;
                return;
            end
        end
    end
end

function path = extract_path(close, start)
% @breif: Extract the path based on the CLOSED set.
    path  = [];               
    closeNum = length(close(:, 1));
    index = 1;
    while 1
        path = [path; close(index, 1:2)];
        if isequal(close(index, 1:2), start)   
            break;
        end
        for i=1:closeNum
            if isequal(close(i, 1:2), close(index, 5:6))
                index = i;
                break;
            end
        end
    end
end
